include/convex_bodies/ellipsoid.h

```diff
@@ -0,0 +1,288 @@
+    return (std::log(x - NT(1)) + log_gamma_function(x - NT(1)));
+}
+
+
+template <class Point, class MT>
+class Ellipsoid{
+private:
+    typedef typename Point::FT NT;
+    typedef typename std::vector<NT>::iterator viterator;
+    typedef Eigen::Matrix<NT, Eigen::Dynamic, 1> VT;
+
+    // representation is (x - c)' A (x - c) <= 1, center is assumed to be origin for now
+    MT A;
+    Point c;
+
+    // MT L;   // LL' = A
+    unsigned int dim;
+
+    // eigen vectors and values
+    VT eigen_values;
+    VT eigen_values_inv;
+    VT eigen_values_inv_sqrt;
+    MT eigen_vecs;
+
+public:
+
+    Ellipsoid(MT& Ain) : A(Ain) {
+        // Eigen::LLT<Eigen::MatrixXd> lltOfA(A); // compute the Cholesky decomposition of Ain
+        // if(lltOfA.info() == Eigen::NumericalIssue) {
+        //     throw std::runtime_error("Possibly non semi-positive definitie matrix!");
+        // }
+        // L = lltOfA.matrixL();
+
+        Eigen::SelfAdjointEigenSolver<MT> eigensolver(A);
+        if (eigensolver.info() != Eigen::Success) {
+            throw std::runtime_error("Eigen solver returned error!");
+        }
+
+        eigen_values = eigensolver.eigenvalues();
+        eigen_vecs = eigensolver.eigenvectors();
+
+        eigen_values_inv = eigen_values.array().inverse().matrix();
+        eigen_values_inv_sqrt = eigen_values_inv.array().sqrt().matrix();
+
+        dim = A.rows();
+        c = Point(dim);
+    }
+
+
+    // Constructor for copula ellipsoid
+    Ellipsoid(std::vector<std::vector<NT> >& Ain) {
+        dim = Ain.size();
+        A.resize(dim, dim);
+        for (unsigned int i = 0; i < Ain.size(); i++) {
+            for (unsigned int j = 0; j < Ain.size(); j++) {
+                A(i,j) = Ain[i][j];
+            }
+        }
+
+        Eigen::SelfAdjointEigenSolver<MT> eigensolver(A);
+        if (eigensolver.info() != Eigen::Success) {
+            throw std::runtime_error("Eigen solver returned error!");
+        }
+
+        eigen_values = eigensolver.eigenvalues();
+        eigen_vecs = eigensolver.eigenvectors();
+
+        eigen_values_inv = eigen_values.array().inverse().matrix();
+        eigen_values_inv_sqrt = eigen_values_inv.array().sqrt().matrix();
+
+        dim = A.rows();
+        c = Point(dim);
+    }
+
+
+    VT eigenvals() const {
+        return eigen_values;
+    }
+
+
+    VT eigenvals_inv() const {
+        return eigen_values_inv;
+    }
+
+
+    VT eigenvals_inv_sqrt() const {
+        return eigen_values_inv_sqrt;
+    }
+
+
+    MT eigenvecs() const {
+        return eigen_vecs;
+    }
+
+
+    unsigned int dimensions() const {
+        return dim;
+    }
+
+
+    void print() {
+        std::cout << "Ellipse is in the form: x' A x <= 1, (center is assumed to be origin always) \n";
+        std::cout << "c = \n" << c.print();
+        std::cout << "A = \n" << A;
+    }
+
+
+
+    NT mat_mult(Point const& p) {
+        VT x = p.getCoefficients();
+
+        if (dim < 15) {
```

Comment:
what happens in `dim = 15`? Does this number come from some benchmark?

Reply:
used this as reference: https://stackoverflow.com/a/63749321

Reply:
hmmm, do you trust the source? There are no benchmarks at all, only a statement "seems to be faster".

Reply:
cool, I will try to benchmark it on my system. Will share the results here

Reply:
benchmarked it here: https://github.com/vaithak/benchmark-symmetric-mat-mult. 
I think we should change `dim < 15` to `dim < 5`, what do you suggest ?

Reply:
thanks for the benchmarks! I think the usage of this part of code is limited since volesti is not focusing in low dimensions (d<=7) so it can be removed for simplicity.

Reply:
ok, removed it 👍🏽